model/src/util/image_hash.rs

```diff
@@ -0,0 +1,775 @@
+    ///
+    /// # Errors
+    ///
+    /// Returns an [`ImageHashParseErrorType::Format`] error type if the
+    /// provided value isn't in a Discord image hash format. Refer to the
+    /// variant's documentation for more details.
+    ///
+    /// Returns an [`ImageHashParseErrorType::Range`] error type if one of the
+    /// hex values is outside of the accepted range. Refer to the variant's
+    /// documentation for more details.
+    fn try_from(value: &str) -> Result<Self, Self::Error> {
+        Self::try_from(value.as_bytes())
+    }
+}
+
+/// Iterator over the [nibbles] of an image hash.
+///
+/// # Examples
+///
+/// Parse an image hash and then iterate over the nibbles:
+///
+/// ```
+/// # fn main() -> Result<(), Box<dyn std::error::Error>> {
+/// use twilight_model::util::ImageHash;
+///
+/// let input = b"1d9811c4cd3782148915c522b02878fc";
+/// let hash = ImageHash::parse(input)?;
+/// let mut nibbles = hash.nibbles();
+///
+/// assert_eq!(Some(b'1'), nibbles.next());
+/// assert_eq!(Some(b'd'), nibbles.next());
+/// assert_eq!(Some(b'c'), nibbles.nth(29));
+/// assert!(nibbles.next().is_none());
+/// # Ok(()) }
+/// ```
+///
+/// [nibbles]: https://en.wikipedia.org/wiki/Nibble
+#[derive(Debug)]
+pub struct Nibbles {
+    /// Current index of the iterator.
+    idx: usize,
+    /// Hash being iterated over.
+    inner: ImageHash,
+}
+
+impl Nibbles {
+    /// Create a new iterator over an image hash, creating nibbles.
+    const fn new(inner: ImageHash) -> Self {
+        Self {
+            idx: usize::MAX,
+            inner,
+        }
+    }
+
+    /// Advance the index in the iterator by the provided amount.
+    fn advance_idx_by(&mut self, by: usize) {
+        self.idx = if self.idx == usize::MAX {
+            0
+        } else {
+            let mut new_idx = self.idx.saturating_add(by);
+
+            if new_idx == usize::MAX {
+                new_idx -= 1;
+            }
+
+            new_idx
+        }
+    }
+
+    /// Parse the byte at the stored index.
+    ///
+    /// # Panics
+    ///
+    /// Panics if [`idx`] is not within the range of `[0, 32)`. Callers must
+    /// ensure that [`idx`] is within that range.
+    ///
+    /// [`idx`]: Self::idx
+    const fn byte(&self) -> Option<u8> {
+        const BITS_IN_HALF_BYTE: u8 = 4;
+        const RIGHT_MASK: u8 = (1 << BITS_IN_HALF_BYTE) - 1;
+
+        if self.idx == usize::MAX || self.idx >= HASH_LEN {
+            return None;
+        }
+
+        let (byte, left) = (self.idx / 2, self.idx % 2 == 0);
+
+        let store = self.inner.bytes[byte];
+
+        let bits = if left {
+            store >> BITS_IN_HALF_BYTE
+        } else {
+            store & RIGHT_MASK
+        };
+
+        Some(Self::nibble(bits))
+    }
+
+    /// Convert 4 bits in a byte integer to a nibble.
+    ///
+    /// Values 0-9 correlate to the ASCII numbers '0' through '9', while values
+    /// 10-15 correlate to the ASCII letters 'a' through 'f'.
```

Comment:
If this is always ASCII, why cant it return a char?

Reply:
this documentation isn't good

Reply:
ah wait i was thinking of something else. it could be a char, but i figured it would be consistent to return bytes when the input to `ImageHash::parse` is also bytes. being a char doesn't really offer any guarantees here or anything

Reply:
Sure, works for me, just wanted to check. Resolving.